subprojects/docs/src/docs/release/notes.md

```diff
@@ -169,9 +116,94 @@ In addition, this behavior can be enabled from the command line for individual b
--->
+### Allow declared reasons for dependency and resolution rules
+
+In complex builds, it can become hard to interpret dependency resolution results and why a [dependency declaration](userguide/customizing_dependency_resolution_behavior.html) or a [rule](userguide/inspecting_dependencies.html#sec:dependency_declaration_reasons) was added to a build script. To improve on this situation, we extended all the corresponding APIs with the capability to define a _reason_ for each declaration or rule. 
+
+These reasons are shown in dependency insight reports and error messages if the corresponding declaration or rule influenced the resolution result. In the future, they will also be shown in build scans.
+
+    dependencies {
+        implementation('org.ow2.asm:asm:6.0') {
+            because 'we require a JDK 9 compatible bytecode generator'
+        }
+    }
+    
+### Dependency constraints for transitive dependencies
+
+With [dependency constraints](userguide/managing_transitive_dependencies.html#sec:dependency_constraints), Gradle adds a mechanism to express constraints over transitive dependencies which are used during dependency resolution. 
+
+    dependencies {
+        implementation 'org.apache.httpcomponents:httpclient'
+    }
+
+    dependencies {
+        constraints {
+            implementation('org.apache.httpcomponents:httpclient:4.5.3') {
+                because 'previous versions have a bug impacting this application'
+            }
+            implementation('commons-codec:commons-codec:1.11') {
+                because 'version 1.9 pulled from httpclient has bugs affecting this application'
+            }
+        }
+    }
+
+In the example, the version of `commons-codec` that is brought in transitively is `1.9`. With the constraint, we express that we need at least `1.11` and Gradle will now pick that version during dependency resolution.
+
+Compared to dependency management rules you can define in the resolution strategy, which are applied after dependency resolution to _fix up_ the result, dependency constraints are declared information that participate in the dependency resolution process.
+This means that all constraints are considered in combination.
+For instance, if another dependency brings in an even higher version of `commons-codec`, Gradle will respect that.
+
+Expressing the same in a traditional dependency substitution rule requires you to repeat part of the dependency resolution process manually – which is expensive and error prone.
+That is, you would have to inspect the version that was selected and implement a decision based on the version:
+
+    resolutionStrategy.eachDependency { DependencyResolveDetails details ->
+        if (details.requested.group == 'commons-codec'
+            && details.requested.name == 'commons-codec'
+            && VersionNumber.parse(details.requested.version) < VersionNumber.parse('1.11')) {
+                details.useVersion '1.11'
+            }
```

Comment:
Sorry this `}` needs to go. I edited that away later (too late it seems).

Reply:
Fixed